File: bxbam.py
import subprocess
import pandas as pd
# import tables    # users must go through downloading this and HDF5 software---put into documentation
import numpy as np
#import numexpr as ne
# also, *must* have Cython installed
import os

bam_path = "/gpfs/commons/home/biederstedte-934/evan_projects/HUMAN_1000Genomes_hs37d5_genomic_RPE1-hTERT_X-33.dupmarked.bam"
bxbam_path = "/gpfs/commons/home/biederstedte-934/evan_projects/HUMAN_1000Genomes_hs37d5_genomic_RPE1-hTERT_X-33.bxbam.h5"

bxbame_name = "HUMAN_1000Genomes_hs37d5_genomic_RPE1-hTERT_X-33.bxbam.h5"

store = pd.HDFStore(bxbame_name)
columns_to_index = [col1, col2]   # list columns to index now, only
col1 = "QNAME"
col2 = "BX"
hdf_key = "default_key"


task = subprocess.Popen("samtools view "+ bam_path, shell=True,  stdout=subprocess.PIPE)
for i, line in enumerate(l.decode(errors='ignore') for l in task.stdout):  # decode binary
    line_split = line.split()           # split line, make list
    mandatory_field = line_split[:11]   # mandatory fields, "values"
    rest_fields = line_split[11:]
    mandatory_field_keys = ["QNAME", "FLAG", "RNAME", "POS", "MAPQ", "CIGAR", "RNEXT", "PNEXT", "TLEN", "SEQ", "QUAL"]
    beginning = '  '.join('{}:{}'.format(k, v) for k, v in zip(mandatory_field_keys, mandatory_field))  # create tsv string
    rest_data = '  '.join(rest_fields)
    total_lines = beginning + rest_data
    # now, parse with awk
    #result = subprocess.check_output(["awk", "-v", "cols=columns.txt", "-f", "bxbam.awk", total_lines])
    result = subprocess.check_output(["awk", "-v", "OFS='\t'", "-f", "bam5.awk", total_lines])
    # result = subprocess.Popen("awk -v OFS='\t' -f bam5.awk "+ total_lines, shell=True,  stdout=subprocess.PIPE)
    store.append(hdf_key, result, data_columns=columns_to_index, index=False)

# index data columns in HDFStore
store.create_table_index(hdf_key, columns=columns_to_index , optlevel=9, kind='full')
store.close()

# awk -v OFS='\t' -f bam5.awk stdout


# It may be best to offer desired order
bxbam_default_columns = ["QNAME", "FLAG", "RNAME", "POS", "MAPQ", "CIGAR", "RNEXT", "PNEXT", "TLEN", "SEQ", "QUAL", "BX"]
# then use ` df = df[bxbam_default_columns]
# if MD? Check whether setting parameter in `create_bxbam()` is best
md_default_columns = ["QNAME", "FLAG", "RNAME", "POS", "MAPQ", "CIGAR", "RNEXT", "PNEXT", "TLEN", "SEQ", "QUAL", "BX", "MD"]


                        
# functions
             
                        
def get_qmates(param1):
    pd.store(bxbam_path).select("QNAME"==param1)

def get_bmates(param1):
    pd.store(bxbam_path).select("BX"==param1)
                        
def get_field(field, param1):
    field = str(field)
    pd.store(bxbam_path).select(field==param1)
                        
def create_bxbam(bam_path, bxbam_path, column1, columns2):
    columns_to_index = [column1, column2]
    # above code
                        
                        
# finish wrapping into an R package `bxbam` that can use the below functions
                        # library(bxbam)

######
#%install.packages("rPython")
#%install.packages("h5")

library(rPython)
                        

# Load python script
python.load("bxbam.py")   # i.e. this script

bxbam_store <-h5file(bxbam)

bam_path <- readline("Enter bam path")
bxbam_path <- readline("Enter desired location/path for bxbam")
           
column1 <- "QNAME"
column2 <- readline("Enter field to index, e.g. BX")   # default option?
                        
get_bmates <- function(reads, "BX" ) {
    pand_func_bmate <- python.exec(bxbamstore.select("BX"))
    python.call("pand_func_bmate", param1)
}

get_qmates <- function(bxbamstore, param1, param2){
    pand_func_qmate <- python.exec(bxbamstore.select("QNAME")
    python.call("pand_func_qmate", param1)
}

get_field <- function(reads, param1) {
    pand_func <- python.exec(bxbamstore.select(param1))
    python.call("pand_func", param1)
}

File: bxbam.R
#############################################################################
## Evan Biederstedt
## New York Genome Center
## ebiederstedt@nygenome.org
##
## Marcin Imielinski
## The Broad Institute of MIT and Harvard / Cancer program.
## marcin@broadinstitute.org
## Weill-Cornell Medical College
## mai9037@med.cornell.edu
## New York Genome Center
## mimielinski@nygenome.org
##
## This program is free software: you can redistribute it and/or modify it
## under the terms of the GNU Lesser General Public License as published by
## the Free Software Foundation, either version 3 of the License, or
## (at your option) any later version.
## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
## GNU General Public License for more details.
## You should have received a copy of the GNU Lesser General Public License
## along with this program.  If not, see <http://www.gnu.org/licenses/>.
##############################################################################

#' @import rPython
#' @import data.table
#' @import Rsamtools
#' @import GenomicRanges
#' @import GenomicAlignments
#' @name load_bxbam
#' @title load_bxbam
#' @description
#'
#' Reads in bxbam object
#'
#' Load the bxbam object (HDF5 format) to be queried via pandas HDFStore() object (PyTables)
#' The default key ('HDF5 group') created using the Python script `create_bxbam.py` is hdf_key = "bam_fields"
#' @param bxbamfile Input HDF5 file to be read, created by script 'create_bxbam.py'
#' @param bam Input bam file
#' @param bami Input bam index file
#' @param hdf_key key defined upon creation of bxBam via 'create_bxbam.py'; default set to hdf_key = "bam_fields"
#' @return Noned
#' @examples
#' load_bxbam("pathname/my_bxbam.h5")
#' @export

load_bxbam <- function(bxbamfile, bam, bami = NULL){     # integreates functionality from BamUtils read.bam
    if (missing(bxbamfile)) stop ("'bxbamfile' is missing; please set correct path to bxbam.h5 object")\
    if (missing(bam)) stop ("'bam' is missing; please set correct path to bam file")
    python.exec("import pandas as pd")
    python.exec("import numpy as np")
    python.exec("import tables")
    python.exec("from tables import *")
    python.assign("bxbamfile", bxbamfile)
    python.exec("store = pd.HDFStore(bxbamfile)")
    python.get("print('Now loaded bxBam file')")
    # python.get("store")   # 'store' variable now saved within Python  ; commented out, Nov. 1 2016
    if (!inherits(bam, 'BamFile'))  # check for bam, bami
    {
        if (is.null(bami))
        {
            if (file.exists(bai <- gsub('.bam$', '.bai', bam)))
            bam = BamFile(bam, bai)
            else if (file.exists(bai <- paste(bam, '.bai', sep = '')))
            bam = BamFile(bam, bai)
            else
            bam = BamFile(bam)
        }
        else
        bam = BamFile(bam, index = bami)
    }
}
#' @name get_bmates
#' @title get_bmates
#' @description
#'
#' Reads R vector of BX strings, outputs GRanges or data.table
#'
#' Function used to output all corresponding bam fields and GRanges via data.table
#' @param reads R vector of BX strings, e.g. bmates_vec <- c("BX1", "BX2", "BX3", "BX4")
#' @param hdf_key key defined upon creation of bxBam via 'create_bxbam.py'; use same hdf_key as bxbam in load_bxbam()
#' @param show_bam    Default FALSE
#' @param as.data.table Return reads in the form of a data.table rather than GRanges/GRangesList
#' @param ignore.indels messes with cigar to read BAM with indels removed. Useful for breakpoint mapping on contigs
#' @param as.grl Return reads as GRangesList. Controls whether \code{get.pairs.grl} does split. Default TRUE
#' @return Corresponding fields in a bam file in data.table, GRanges, or GRangesList
#' @export
get_bmates <- function(reads, bxbam_key = "bam_fields", pairs.grl = TRUE, ignore.indels= FALSE, as.data.table = FALSE){   # show_bam = FALSE,
    if (is.null(reads))
    {
        stop('Must provide non empty read vector')
    }
    python.exec("import pandas as pd")
    python.exec("import numpy as np")
    python.exec("import tables")
    python.exec("from tables import *")
    python.assign("reads", reads)
    pand_func_bmate <- python.exec("df = store.select(bxbam_key, where='BX in @reads')")
    df <- python.get("df")    #!!!!!!!!!!!!!! is this crappy R code/style?
    df <- data.table(df)   # outputs data.table of bam rows   #!!!!!!!!!!!!!! is this crappy R code/style?
    
    # we now have data.frame of BAM fields (in the format as out in bamUtils
    #
    
    ## faster CIGAR string parsing with vectorization and data tables
    if (verbose) {
        print(Sys.time() - now)
        print('filling pos2 from cigar')
    }
    if (ignore.indels) {
        cigar <- gsub('[0-9]+D', '', gsub('([0-9]+)I', '\\1M', out$cigar))  ## Remove deletions, turn insertions to matches
        cig <- splitCigar(cigar)
        torun=sapply(cig, function(y) any(duplicated((y[[1]][y[[1]]==M]))))
        M <- charToRaw('M')
        new.cigar <- sapply(cig[torun], function(y) {
            lets <- y[[1]][!duplicated(y[[1]])]
            vals <- y[[2]][!duplicated(y[[1]])]
            vals[lets==M] <- sum(y[[2]][y[[1]]==M])
            lets <- strsplit(rawToChar(lets), '')[[1]]
            paste(as.vector(t(matrix(c(vals, lets), nrow=length(vals), ncol=length(lets)))), collapse='')
        })
        out$cigar[torun] <- new.cigar
    }
    cigs <- countCigar(out$cigar)
    out$pos2 <- out$pos + cigs[, "M"]
    
    if (verbose) {
        print(Sys.time() - now)
        print('fixing seqdata')
    }
    out$qwidth = nchar(out$seq)
    unm = is.na(out$pos)
    if (any(unm))
    {
        out$pos[unm] = 1
        out$pos2[unm] = 0
        out$strand[unm] = '*'
    }
    gr.fields = c('rname', 'strand', 'pos', 'pos2');
    vals = out[, setdiff(names(out), gr.fields)]
    
    if (!as.data.table) {
        out <- GRanges(out$rname, IRanges(out$pos, pmax(0, out$pos2-1)), strand = out$strand, seqlengths = seqlengths(intervals))
        values(out) <- vals;
    } else {
        out <- data.table(seqnames=out$rname, start=out$pos, end= pmax(out$pos2-1, 0), strand=out$strand)
        val <- data.table(vals)
        out <- cbind(out, val)
    }
    #out$uname = paste(out$qname, ifelse(bamflag(out$flag)[, 'isFirstMateRead'], '_r1', '_r2'), sep = '')
    }
    else {
    if (!as.data.table)
    return(GRanges(seqlengths = seqlengths(intervals)))
    else
    return(data.table())
}

if (verbose)
{
    if (as.data.table)
    cat(sprintf('Extracted %s reads\n', nrow(out)))
    else
    cat(sprintf('Extracted %s reads\n', length(out)))
    print(Sys.time() - now)
}

if (pairs.grl)
{
    if (verbose)
    cat('Pairing reads\n')
    out <- get.pairs.grl(out, as.grl=as.grl)
    if (verbose)
    {
        cat('done\n')
        print(Sys.time() - now)
    }
    if (as.grl && !as.data.table) {
        names(out) = NULL;
        values(out)$col = 'gray';
        values(out)$border = 'gray';
    }
}

return(out)
}


}

## Ditto with above----FIX!!!!!

#' @name get_qmates
#' @title get_qmates
#' @description
#'
#' Reads R vector of QNAME strings, outputs GRanges or data.table
#'
#' Function used to output all corresponding bam fields and GRanges via data.table
#' @param reads R vector of QNAME strings, e.g. bmates_vec <- c("QNAME1", "QNAME2", "QNAME3", "QNAME4")
#' @param hdf_key key defined upon creation of bxBam via 'create_bxbam.py'; use same hdf_key as bxbam in load_bxbam()
#' @param show_bam    Default False
#' @return Corresponding fields in a bam file in data.table, GRanges, or GRangesList
#' @export
get_qmates <- function(reads, bxbam_key = "bam_fields", show_bam = FALSE){
    python.exec("import pandas as pd")
    python.exec("import numpy as np")
    python.exec("import tables")
    python.exec("from tables import *")
    python.assign("reads", reads)
    pand_func_bmate <- python.exec("df = store.select(bxbam_key, where='QNAME in @reads'")
    df <- python.get("df")
    df <- data.table(df)
    return(df)
}

